File: packages/sdk/src/commands/detect.ts
import { type DetectedHost, detectHosts } from '../lib/detect.js';
import { type EcosystemReport, detectEcosystem, renderEcosystemLine } from '../lib/ecosystem.js';

/** Result of `detect`: the agent hosts available to this user, ranked, and the
 * dependency ecosystem of the repository itself. */
export interface DetectResult {
   hosts: DetectedHost[];
   ecosystem: EcosystemReport;
}

export function detectLayer(root: string): DetectResult {
   return { hosts: detectHosts({ root }), ecosystem: detectEcosystem(root) };
}

/** Human-readable detection report. */
export function renderDetect(result: DetectResult): string {
   const { hosts, ecosystem } = result;
   const ecoLine = renderEcosystemLine(ecosystem);
   if (hosts.length === 0) {
      return `No coding-agent hosts detected. Leji works without one; the onboarding brief still guides any agent you point at it.\n\n${ecoLine}`;
   }
   const lines = ['Detected agent hosts (strongest signal first):'];
   for (const h of hosts) {
      const signals = [h.onPath && 'binary on PATH', h.inRepo && 'config in repo', h.userConfig && 'user config']
         .filter(Boolean)
         .join(', ');
      const adapter = h.adapter ? `adapter ${h.adapter}` : 'directory-style adapter (wiring deferred)';
      lines.push(`   ${h.strength.padEnd(16)} ${h.name}: ${signals}; ${adapter}`);
   }
   // One line about the repository's own ecosystem: what would declare and run the
   // CLI here. The full offer block belongs to init/adopt, which can act on it.
   lines.push('', ecoLine);
   // `--agent` names the host Leji launches, and only claude-code and codex accept
   // an inline prompt; suggesting `--agent <name>` for every detected host offered
   // a command the flag rejects.
   lines.push(
      '',
      '--agent takes a launchable host, claude-code or codex: leji init --agent claude-code, leji start --agent codex.',
      'Any other host above enters the layer through its vendor-file redirect.',
   );
   return lines.join('\n');
}

File: packages/sdk-go/internal/commands/detect/detect.go
// Package detectcmd renders the human-readable agent-host detection report.
package detectcmd

import (
	"strings"

	"github.com/leji-org/leji/packages/sdk-go/internal/detect"
	"github.com/leji-org/leji/packages/sdk-go/internal/ecosystem"
)

// DetectLayer returns the agent hosts available to this user, ranked.
func DetectLayer(root string) []detect.DetectedHost {
	return detect.DetectHosts(detect.Options{Root: root})
}

func RenderDetect(hosts []detect.DetectedHost, eco ecosystem.Report) string {
	ecoLine := ecosystem.RenderLine(eco)
	if len(hosts) == 0 {
		return "No coding-agent hosts detected. Leji works without one; the onboarding brief still guides any agent you point at it." +
			"\n\n" + ecoLine
	}
	lines := []string{"Detected agent hosts (strongest signal first):"}
	for _, h := range hosts {
		var sig []string
		if h.OnPath {
			sig = append(sig, "binary on PATH")
		}
		if h.InRepo {
			sig = append(sig, "config in repo")
		}
		if h.UserConfig {
			sig = append(sig, "user config")
		}
		signals := strings.Join(sig, ", ")
		adapter := "directory-style adapter (wiring deferred)"
		if h.Adapter != "" {
			adapter = "adapter " + h.Adapter
		}
		lines = append(lines, "   "+padEnd(string(h.Strength), 16)+" "+h.Name+": "+signals+"; "+adapter)
	}
	// One line about the repository's own ecosystem: what would declare and run the
	// CLI here. The full offer block belongs to init/adopt, which can act on it.
	lines = append(lines, "", ecoLine)
	// --agent names the host Leji launches, and only claude-code and codex accept
	// an inline prompt; suggesting `--agent <name>` for every detected host offered
	// a command the flag rejects.
	lines = append(lines, "",
		"--agent takes a launchable host, claude-code or codex: leji init --agent claude-code, leji start --agent codex.",
		"Any other host above enters the layer through its vendor-file redirect.")
	return strings.Join(lines, "\n")
}

// padEnd right-pads s with spaces to at least width.
func padEnd(s string, width int) string {
	if len(s) >= width {
		return s
	}
	return s + strings.Repeat(" ", width-len(s))
}
